File: vb/high_perf.vb
Imports System.Net
Imports System.Threading.Tasks

Class Client
    Inherits WebClient
    Const username As String = "lum-customer-CUSTOMER-zone-YOURZONE"
    Const password As String = "YOURPASS"
    Const port = 22225
    Public session_id As String = New Random().Next().ToString()

    Public Sub New(Optional country As String = Nothing)
        Me.Proxy = New WebProxy("zproxy.luminati.io", port)
        Dim login = username &
            If(country IsNot Nothing, "-country-" & country, "") &
            "-session-" & session_id
        Me.Proxy.Credentials = New NetworkCredential(login, password)
    End Sub

    Protected Overrides Function GetWebRequest(address As Uri) As WebRequest
        Dim request = MyBase.GetWebRequest(address)
        request.ConnectionGroupName = session_id
        Return request
    End Function
End Class

Module Example
    Sub Main()
        ' Example of how to run 100 parallel sessions (100 different Exit Node
        ' IPs) by using 100 threads, each with a unique session_id.
        Parallel.For(0, 100,
        Sub(i)
            Dim session as New Client()
            Console.WriteLine("Performing request(s) from session #{0}", i)
            ' Put complete scraping sequence that should be done in one session (from one IP):
            Console.WriteLine(session.DownloadString("http://lumtest.com/myip.json"))
            ' session.DownloadString(...second request...);
        End Sub)
    End Sub
End Module

File: csharp/high_perf.cs



using System;
using System.Threading;
using System.Net;
using System.Collections.Generic;

class Client : WebClient
{
    public static string username = "lum-customer-CUSTOMER-zone-YOURZONE";
    public static string password = "YOURPASS";
    public static int port = 22225;
    public static int max_failures = 3;
    public string session_id;
    public string login;
    public string country;
    public int fail_count;
    public int n_req_for_exit_node;
    public Random rng;
    public HashSet<ServicePoint> service_points;

    public Client(string country = null)
    {
        this.country = country;
        rng = new Random();
        service_points = new HashSet<ServicePoint>();
        switch_session_id();
    }

    public void switch_session_id()
    {
        clean_connection_pool();
        session_id = rng.Next().ToString();
        n_req_for_exit_node = 0;
        update_super_proxy();
    }

    public void update_super_proxy()
    {
        Proxy = new WebProxy("session-"+session_id+".zproxy.luminati.io", port);
        login = username+(country != null ? "-country-"+country : "")
            +"-session-"+session_id;
        Proxy.Credentials = new NetworkCredential(login, password);
    }

    public void clean_connection_pool()
    {
        foreach (ServicePoint sp in service_points)
            sp.CloseConnectionGroup(login);
        service_points.Clear();
    }

    public bool have_good_super_proxy()
    {
        return fail_count < max_failures;
    }

    public void handle_response(WebException e = null) {
        if (e != null && should_switch_exit_node((HttpWebResponse)e.Response))
        {
            switch_session_id();
            fail_count++;
            return;
        }
        // success or other client/website error like 404...
        n_req_for_exit_node++;
        fail_count = 0;
    }

    public bool should_switch_exit_node(HttpWebResponse response)
    {
        return response == null ||
            status_code_requires_exit_node_switch((int)response.StatusCode);
    }

    public bool status_code_requires_exit_node_switch(int code)
    {
        return code == 403 || code == 429 || code==502 || code == 503;
    }

    protected override WebRequest GetWebRequest(Uri address)
    {
        var request = base.GetWebRequest(address) as HttpWebRequest;
        request.ConnectionGroupName = login;
        request.PreAuthenticate = true;
        return request;
    }

    protected override WebResponse GetWebResponse(WebRequest request)
    {
        var response = base.GetWebResponse(request);
        ServicePoint sp = ((HttpWebRequest)request).ServicePoint;
        service_points.Add(sp);
        return response;
    }
}

class Example
{
    public static int n_parallel_exit_nodes = 100;
    public static int n_total_req = 1000;
    public static int switch_ip_every_n_req = 20;
    public static int at_req = 0;

    static void Main()
    {
        ServicePointManager.DefaultConnectionLimit = int.MaxValue;
        for (var i = 0; i < n_parallel_exit_nodes; i++)
        {
            var t = new Thread(new ThreadStart(Run));
            t.Name = ""+i;
            t.Start();
        }
    }

    static void Run()
    {
        var client = new Client();
        while (Interlocked.Increment(ref at_req) <= n_total_req)
        {
            if (!client.have_good_super_proxy())
                client.switch_session_id();
            if (client.n_req_for_exit_node == switch_ip_every_n_req)
                client.switch_session_id();
            try {
                Console.WriteLine(client.DownloadString("http://lumtest.com/myip.json"));
                client.handle_response();
            } catch (WebException e) {
                Console.WriteLine(e.Message);
                client.handle_response(e);
            }
        }
        client.clean_connection_pool();
        client.Dispose();
    }
}

